### PR TITLE
Sprites

### PR DESCRIPTION
**Pull request by davidmanzanares**
Adds https://github.com/CartoDB/carto-vl/issues/266

Example: http://127.0.0.1:8080/examples/editor/#eyJhIjoidHJhZmZpY19hY2NpZGVudHNfY29weSIsImIiOiIiLCJjIjoibWFtYXRhYWtlbGxhIiwiZCI6Imh0dHBzOi8ve3VzZXJ9LmNhcnRvLmNvbSIsImUiOiJ3aWR0aDogICAkY291bnQvMlxuY29sb3I6IG9wYWNpdHkoIHJhbXAobGluZWFyKCRjb3VudCwgMCwxMjApLCBSZWRPciksICRjb3VudC8yMClcbnN5bWJvbFBsYWNlbWVudDogcGxhY2VtZW50KHNpbigkY291bnQqbm93KCkpICwgY29zKCRjb3VudCpub3coKSkpXG5zeW1ib2w6IHNwcml0ZSgnLi4vZXhwcmVzc2lvbnMvbWFya2VyLnN2ZycpXG5cbiAiLCJmIjp7ImxuZyI6LTEwNC45NTkxMjEwNzMzMzIzMSwibGF0IjozOS43MTA4OTc3NzM4NjE4fSwiZyI6MTAuODA1NTQxOTIyNzc4NzUzfQ==

To be done on this PR:
- [x] `symbol:`
- [x] `symbolPlacement`
- [x] Support multi-sprite selection based on category
- [x] complete the refactor of the shader compilation process to remove unnecessary classes and allow for the generation of one shader from multiple expressions
- [x] Add render tests
- [x] Fix the re-render issue on sprites
- [x] Fix the render tests scripts to avoid leaking the http-server
- [x] Review the entire PR (myself included)
- [x] Add reference docs

To be done on future PRs:
- Add a guide
- Discuss cross-origin problems and possible solutions/workarounds
- Discuss `color` and `strokeColor` semantics
- Maybe, provide some default sprites
- Improve interactivity for sprites

### CODE REVIEW

File: src/api/viz.js [hunk 1]
```diff
@@ -267,45 +264,39 @@ export default class Viz {
...
+        this.symbolPlacement._bind(metadata);
+        if (!this.symbol._default) {
+            this.symbolShader = compileShader(gl, symbolizerGLSL, {
+                symbol: this.symbol,
+                symbolPlacement: this.symbolPlacement
+            });
+        }
+
+        Object.values(this.variables).map(v => {
+            v._bind(metadata);
```
Comment (davidmanzanares) [inline src/api/viz.js]: This may be not needed

File: src/core/shaders/index.js [hunk 2]
```diff
@@ -17,10 +63,55 @@ const renderer = {
     createLineShader: gl => new LineShader(gl),
 };
 
+// TODO remove class nonsense
```
Comment (davidmanzanares) [inline src/core/shaders/index.js]: ?

File: src/core/viz/expressions/ramp.js [hunk 3]
```diff
@@ -144,12 +160,20 @@ export default class Ramp extends BaseExpression {
...
+            super._postShaderCompile(program, gl);
+            return;
+        }
         this.input._postShaderCompile(program, gl);
         this._getBinding(program).texLoc = gl.getUniformLocation(program, `texRamp${this._uid}`);
         this._getBinding(program).keyMinLoc = gl.getUniformLocation(program, `keyMin${this._uid}`);
         this._getBinding(program).keyWidthLoc = gl.getUniformLocation(program, `keyWidth${this._uid}`);
     }
     _computeTextureIfNeeded() {
+        if (this.palette.type == 'sprites') {
```
Comment (davidmanzanares) [inline src/core/viz/expressions/ramp.js]: Not needed

File: src/core/viz/expressions/sprite.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        this._url = url;
+    }
+    _fetch() {
+        return new Promise((resolve, reject) => {
+            this.image = new Image();
+            this.image.onload = () => {
+                this.image = getCanvasFromImage(this.image);
+                this.ready = true;
+                resolve();
+                console.log('resolved!');
```
Comment (davidmanzanares) [inline src/core/viz/expressions/sprite.js]: !

File: src/core/viz/expressions/sprite.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            };
+            this.image.onerror = reject;
+            this.image.src = this._url;
+        });
+    }
+    _compile(meta) {
+        super._compile(meta);
```
Comment (davidmanzanares) [inline src/core/viz/expressions/sprite.js]: not needed

File: src/core/viz/expressions/sprite.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    }
+    _free(gl) {
+        gl.deleteTexture(this.texture);
```
Comment (davidmanzanares) [inline src/core/viz/expressions/sprite.js]: if...

File: src/core/viz/expressions/sprite.js [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+            inline:
+                `texture2D(texSprite${this._uid}, spriteUV).rgba`
+        };
+    }
+
+    _postShaderCompile(program, gl) {
+        this._getBinding(program)._texLoc = gl.getUniformLocation(program, `texSprite${this._uid}`);
+    }
+    _preDraw(program, drawMetadata, gl) {
+        if (!this.init && this.ready) {
```
Comment (davidmanzanares) [inline src/core/viz/expressions/sprite.js]: if this.texture...

File: src/core/viz/expressions/sprite.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+            drawMetadata.freeTexUnit++;
+            console.log('DRAW!');
+        }
+    }
+    // TODO eval
+}
+
+function getCanvasFromImage(img) {
+    const canvasSize = 256;
+    const canvas = document.createElement('canvas');
```
Comment (davidmanzanares) [inline src/core/viz/expressions/sprite.js]: performance, re use

File: src/core/viz/expressions/sprites.js [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        this.numSprites = sprites.length;
+        this.type = 'sprites';
+    }
+    _applyToShaderSource() {
+        return {
+            preface: this._prefaceCode(`
+        uniform sampler2D atlas${this._uid};
+
+        vec4 atlas${this._uid}Fn(vec2 spriteUV, float cat){
+            /*if (true){
```
Comment (davidmanzanares) [inline src/core/viz/expressions/sprites.js]: remove comment

**Comment by elenatorro** [comment]
Hi :) we have seen an error in the `sprites` expression when using a ramp of sprites to set the symbol markers:

```
width: 15
symbol: ramp($amount, sprites(sprite('../expressions/marker.svg'), sprite('../expressions/marker2.svg')))
``` 
captura de pantalla 2018-06-04 a las 13 28 07

When using a linear expression, if the higher limit is not big enough, the svgs also look like clipped.

1. min 0 max 100 👎 
captura de pantalla 2018-06-04 a las 15 21 26

2. min 0 max 10k 👍 
captura de pantalla 2018-06-04 a las 15 23 53

File: src/core/shaders/styler.js [hunk 10]
```diff
@@ -0,0 +1,29 @@
+//TODO Discuss size scaling constant, maybe we need to remap using an exponential map
+
```
Comment (davidmanzanares) [inline src/core/shaders/styler.js]: Review if this code is redundant.

File: src/core/shaders/symbolizer.js [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+uniform sampler2D filterTex;
+//TODO order bucket texture
+
+varying highp vec2 featureIDVar;
+varying highp vec4 color;
+varying highp float fillScale;
+
+// From [0.,1.] in exponential-like form to pixels in [0.,255.]
+float decodeWidth(float x){
+    x*=255.;
```
Comment (davidmanzanares) [inline src/core/shaders/symbolizer.js]: Be consistent!

File: src/core/viz/expressions/placement.js [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+ * Placement.
+ *
+ * @param {number} x - first numeric expression
+ * @param {number} y - second numeric expression
+ * @return {carto.expressions.Base} Numeric expression
+ *
+ * @example <caption>Creating a number expression.</caption>
+ * const s = carto.expressions;
+ * const viz = new carto.Viz({
+ *   width: s.number(15);  // Equivalent to `width: 15`
```
Comment (davidmanzanares) [inline src/core/viz/expressions/placement.js]: Fix!

File: src/core/viz/expressions/sprite.js (hunk 8 above)
Comment (davidmanzanares) [reply src/core/viz/expressions/sprite.js]: solved by @elenatorro

File: src/core/viz/expressions/sprites.js [hunk 13]
```diff
@@ -0,0 +1,80 @@
+import Base from './base';
+
```
Comment (davidmanzanares) [inline src/core/viz/expressions/sprites.js]: docs!

File: src/core/viz/expressions/sprites.js [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+export default class Sprites extends Base {
+    constructor(...sprites) {
```
Comment (davidmanzanares) [inline src/core/viz/expressions/sprites.js]: use an array

File: src/core/viz/parser.js [hunk 15]
```diff
@@ -15,6 +15,8 @@ Object.keys(functions)
     .map(name => { lowerCaseFunctions[name.toLocaleLowerCase()] = functions[name]; });
 lowerCaseFunctions.true = functions.TRUE;
 lowerCaseFunctions.false = functions.FALSE;
+lowerCaseFunctions.center = functions.CENTER;
```
Comment (davidmanzanares) [inline src/core/viz/parser.js]: placement center?

File: examples/expressions/sprite-ramp.html [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+      user: 'cartogl',
+      apiKey: 'YOUR_API_KEY'
+    });
+
+    const source = new carto.source.Dataset('ne_10m_populated_places_simple');
+    const viz = new carto.Viz(`
+    width: sqrt($pop_max) * 0.03
+    color: ramp(linear($pop_max, 0, 10000000), prism)
+    symbol: ramp(buckets($pop_max, [1000000]), sprites(sprite('./marker.svg'), sprite('./marker2.svg')))
+    symbolPlacement: bottom
```
[the remark this answers is not in the document]
Comment (davidmanzanares) [reply examples/expressions/sprite-ramp.html]: Fixed

File: src/core/viz/expressions/placement.js (hunk 12 above)
Comment (davidmanzanares) [reply src/core/viz/expressions/placement.js]: Fixed

File: src/core/viz/parser.js (hunk 15 above)
Comment (davidmanzanares) [reply src/core/viz/parser.js]: Fixed (`align_center`)

File: src/core/viz/expressions/sprite.js [hunk 17]
```diff
@@ -0,0 +1,90 @@
+import Base from './base';
+
+export default class Sprite extends Base {
+    constructor(url) {
```
Comment (elenatorro) [inline src/core/viz/expressions/sprite.js]: check typing

File: src/core/shaders/styler.js (hunk 10 above)
Comment (davidmanzanares) [reply src/core/shaders/styler.js]: Fixed

**Comment by elenatorro** [comment]
LGTM

**Pull request merged**